{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "30ed91c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import sys\n",
    "import re\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2361d9b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "if len(sys.argv) < 2:\n",
    "    print('Usage: ' + sys.argv[0] + '<input_file>' + '<output_file>')\n",
    "    sys.exit(1)\n",
    "else:\n",
    "    print('Input file: ' + sys.argv[1])\n",
    "    print('Output file: ' + sys.argv[2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "b982e2ef",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "##fileformat=VCFv4.1\n",
      "\n",
      "##fileDate=20230517\n",
      "\n",
      "##source=freeBayes v0.9.21\n",
      "\n",
      "##reference=k55hyb_GUR2.fasta\n",
      "\n",
      "##phasing=none\n",
      "\n",
      "##commandline=\"freebayes -f k55hyb_GUR2.fasta -C 1 -F 0.01 -i -4 -b allgmref_trimmfastp_paired.wg.srtcrd.fix.mkdp_idfix.L1000.bam --pooled-discrete --cnv-map chromploidy.bed -p 2 --populations populations.file --failed-alleles allpools_gmref_failedalleles.txt --use-best-n-alleles 2 --vcf allpools_gmref_C1F001_wdup.vcf\"\n",
      "\n",
      "##INFO=<ID=NS,Number=1,Type=Integer,Description=\"Number of samples with data\">\n",
      "\n",
      "##INFO=<ID=DP,Number=1,Type=Integer,Description=\"Total read depth at the locus\">\n",
      "\n",
      "##INFO=<ID=DPB,Number=1,Type=Float,Description=\"Total read depth per bp at the locus; bases in reads overlapping / bases in haplotype\">\n",
      "\n",
      "##INFO=<ID=AC,Number=A,Type=Integer,Description=\"Total number of alternate alleles in called genotypes\">\n",
      "\n",
      "##INFO=<ID=AN,Number=1,Type=Integer,Description=\"Total number of alleles in called genotypes\">\n",
      "\n",
      "##INFO=<ID=AF,Number=A,Type=Float,Description=\"Estimated allele frequency in the range (0,1]\">\n",
      "\n",
      "##INFO=<ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "\n",
      "##INFO=<ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "\n",
      "##INFO=<ID=PRO,Number=1,Type=Float,Description=\"Reference allele observation count, with partial observations recorded fractionally\">\n",
      "\n",
      "##INFO=<ID=PAO,Number=A,Type=Float,Description=\"Alternate allele observations, with partial observations recorded fractionally\">\n",
      "\n",
      "##INFO=<ID=QR,Number=1,Type=Integer,Description=\"Reference allele quality sum in phred\">\n",
      "\n",
      "##INFO=<ID=QA,Number=A,Type=Integer,Description=\"Alternate allele quality sum in phred\">\n",
      "\n",
      "##INFO=<ID=PQR,Number=1,Type=Float,Description=\"Reference allele quality sum in phred for partial observations\">\n",
      "\n",
      "##INFO=<ID=PQA,Number=A,Type=Float,Description=\"Alternate allele quality sum in phred for partial observations\">\n",
      "\n",
      "##INFO=<ID=SRF,Number=1,Type=Integer,Description=\"Number of reference observations on the forward strand\">\n",
      "\n",
      "##INFO=<ID=SRR,Number=1,Type=Integer,Description=\"Number of reference observations on the reverse strand\">\n",
      "\n",
      "##INFO=<ID=SAF,Number=A,Type=Integer,Description=\"Number of alternate observations on the forward strand\">\n",
      "\n",
      "##INFO=<ID=SAR,Number=A,Type=Integer,Description=\"Number of alternate observations on the reverse strand\">\n",
      "\n",
      "##INFO=<ID=SRP,Number=1,Type=Float,Description=\"Strand balance probability for the reference allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SRF and SRR given E(SRF/SRR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "\n",
      "##INFO=<ID=SAP,Number=A,Type=Float,Description=\"Strand balance probability for the alternate allele: Phred-scaled upper-bounds estimate of the probability of observing the deviation between SAF and SAR given E(SAF/SAR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "\n",
      "##INFO=<ID=AB,Number=A,Type=Float,Description=\"Allele balance at heterozygous sites: a number between 0 and 1 representing the ratio of reads showing the reference allele to all reads, considering only reads from individuals called as heterozygous\">\n",
      "\n",
      "##INFO=<ID=ABP,Number=A,Type=Float,Description=\"Allele balance probability at heterozygous sites: Phred-scaled upper-bounds estimate of the probability of observing the deviation between ABR and ABA given E(ABR/ABA) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "\n",
      "##INFO=<ID=RUN,Number=A,Type=Integer,Description=\"Run length: the number of consecutive repeats of the alternate allele in the reference genome\">\n",
      "\n",
      "##INFO=<ID=RPP,Number=A,Type=Float,Description=\"Read Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "\n",
      "##INFO=<ID=RPPR,Number=1,Type=Float,Description=\"Read Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between RPL and RPR given E(RPL/RPR) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "\n",
      "##INFO=<ID=RPL,Number=A,Type=Float,Description=\"Reads Placed Left: number of reads supporting the alternate balanced to the left (5') of the alternate allele\">\n",
      "\n",
      "##INFO=<ID=RPR,Number=A,Type=Float,Description=\"Reads Placed Right: number of reads supporting the alternate balanced to the right (3') of the alternate allele\">\n",
      "\n",
      "##INFO=<ID=EPP,Number=A,Type=Float,Description=\"End Placement Probability: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "\n",
      "##INFO=<ID=EPPR,Number=1,Type=Float,Description=\"End Placement Probability for reference observations: Phred-scaled upper-bounds estimate of the probability of observing the deviation between EL and ER given E(EL/ER) ~ 0.5, derived using Hoeffding's inequality\">\n",
      "\n",
      "##INFO=<ID=DPRA,Number=A,Type=Float,Description=\"Alternate allele depth ratio.  Ratio between depth in samples with each called alternate allele and those without.\">\n",
      "\n",
      "##INFO=<ID=ODDS,Number=1,Type=Float,Description=\"The log odds ratio of the best genotype combination to the second-best.\">\n",
      "\n",
      "##INFO=<ID=GTI,Number=1,Type=Integer,Description=\"Number of genotyping iterations required to reach convergence or bailout.\">\n",
      "\n",
      "##INFO=<ID=TYPE,Number=A,Type=String,Description=\"The type of allele, either snp, mnp, ins, del, or complex.\">\n",
      "\n",
      "##INFO=<ID=CIGAR,Number=A,Type=String,Description=\"The extended CIGAR representation of each alternate allele, with the exception that '=' is replaced by 'M' to ease VCF parsing.  Note that INDEL alleles do not have the first matched base (which is provided by default, per the spec) referred to by the CIGAR.\">\n",
      "\n",
      "##INFO=<ID=NUMALT,Number=1,Type=Integer,Description=\"Number of unique non-reference alleles in called genotypes at this position.\">\n",
      "\n",
      "##INFO=<ID=MEANALT,Number=A,Type=Float,Description=\"Mean number of unique non-reference allele observations per sample with the corresponding alternate alleles.\">\n",
      "\n",
      "##INFO=<ID=LEN,Number=A,Type=Integer,Description=\"allele length\">\n",
      "\n",
      "##INFO=<ID=MQM,Number=A,Type=Float,Description=\"Mean mapping quality of observed alternate alleles\">\n",
      "\n",
      "##INFO=<ID=MQMR,Number=1,Type=Float,Description=\"Mean mapping quality of observed reference alleles\">\n",
      "\n",
      "##INFO=<ID=PAIRED,Number=A,Type=Float,Description=\"Proportion of observed alternate alleles which are supported by properly paired read fragments\">\n",
      "\n",
      "##INFO=<ID=PAIREDR,Number=1,Type=Float,Description=\"Proportion of observed reference alleles which are supported by properly paired read fragments\">\n",
      "\n",
      "##INFO=<ID=technology.illumina,Number=A,Type=Float,Description=\"Fraction of observations supporting the alternate observed in reads from illumina\">\n",
      "\n",
      "##FORMAT=<ID=GT,Number=1,Type=String,Description=\"Genotype\">\n",
      "\n",
      "##FORMAT=<ID=GQ,Number=1,Type=Float,Description=\"Genotype Quality, the Phred-scaled marginal (or unconditional) probability of the called genotype\">\n",
      "\n",
      "##FORMAT=<ID=GL,Number=G,Type=Float,Description=\"Genotype Likelihood, log10-scaled likelihoods of the data given the called genotype for each possible genotype generated from the reference and alternate alleles given the sample ploidy\">\n",
      "\n",
      "##FORMAT=<ID=DP,Number=1,Type=Integer,Description=\"Read Depth\">\n",
      "\n",
      "##FORMAT=<ID=RO,Number=1,Type=Integer,Description=\"Reference allele observation count\">\n",
      "\n",
      "##FORMAT=<ID=QR,Number=1,Type=Integer,Description=\"Sum of quality of the reference observations\">\n",
      "\n",
      "##FORMAT=<ID=AO,Number=A,Type=Integer,Description=\"Alternate allele observation count\">\n",
      "\n",
      "##FORMAT=<ID=QA,Number=A,Type=Integer,Description=\"Sum of quality of the alternate observations\">\n",
      "\n",
      "1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1\n",
      "1\n",
      "0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0\n",
      "1\n",
      "0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0\n",
      "1\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0\n",
      "1\n",
      "0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0\n",
      "1\n",
      "0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0\n",
      "1\n",
      "0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0/0\n",
      "1\n",
      "1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1/1\n",
      "1\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[20], line 51\u001b[0m\n\u001b[1;32m     34\u001b[0m             \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mset\u001b[39m(alleles)))\n\u001b[1;32m     37\u001b[0m         \u001b[38;5;66;03m# if 0/1 AD takes the form of x,y where x is reference allele count and y alternative allele count\u001b[39;00m\n\u001b[1;32m     38\u001b[0m         \u001b[38;5;66;03m# check if genotype is 0/1\u001b[39;00m\n\u001b[1;32m     39\u001b[0m         \n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     49\u001b[0m \u001b[38;5;66;03m#         newline = re.sub(oldformat, newformat, line)\u001b[39;00m\n\u001b[1;32m     50\u001b[0m \u001b[38;5;66;03m#         output.write(newformat)\u001b[39;00m\n\u001b[0;32m---> 51\u001b[0m         \u001b[43mtime\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msleep\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m0.5\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# create file from output argument\n",
    "# X \n",
    "# Number=A means there is integers, variable tuples (can be one or multiple delimited by comma)\n",
    "new_annotation = '##FORMAT=<ID=AD,Number=A,Type=Integer,Description=\"alternate allele depth\">'\n",
    "id_in_format = 'AD'\n",
    "\n",
    "# read input vcf\n",
    "for line in open('./allpools_gmref_C1F001_wdup.vcf','r'):\n",
    "    if line.startswith('##') == True: # headers already present\n",
    "        # output.write('%s\\n'%(line)) # print lines straight into the file\n",
    "        print(line)\n",
    "        if line.startswith('#CHROM') == True:\n",
    "    #         output.write('%s\\n%s\\n'%(new_annotation,line)) if new_annotation != str(line) else output.write('%s\\n'%(line))\n",
    "            print('%s\\n%s\\n'%(new_annotation,line)) if new_annotation != str(line) else print('%s\\n'%(line))\n",
    "    else:\n",
    "        # now dealing with the variants and their annotation\n",
    "        # get the format field of the line\n",
    "        oldformat = line.split('\\t')[8]\n",
    "        # add the new AD flag\n",
    "        newformat = '%s:AD'%(processing)\n",
    "        \n",
    "        # improve the following by indexing lines, assinging cases (allele num and type) then replace and print by blocks\n",
    "        # for annotating allele depth\n",
    "        # if 0/0 == annotation is reference count (single number)\n",
    "        # check if genotype is  0/0\n",
    "        genotype = line.split('\\t')[9].split(':')[0]\n",
    "        alleles = re.findall(r'\\d+', genotype)\n",
    "        \n",
    "        \n",
    "        # AC total numb of alternate alleles in called gens\n",
    "        # AN total num of alleles in called gens (alternate plus ref)\n",
    "        if len(set(alleles)) == 1 :\n",
    "            print(genotype)\n",
    "            print(len(set(alleles)))\n",
    "\n",
    "        \n",
    "        # if 0/1 AD takes the form of x,y where x is reference allele count and y alternate allele count\n",
    "        # check if genotype is 0/1\n",
    "        \n",
    "        # if 1/2 AD takes form y,z where y is alternate allele one count and z is allele two count\n",
    "        # check if genotype is 1/2\n",
    "        \n",
    "        \n",
    "#         oldformvalue = line.split('\\t')[9]\n",
    "#         adformat = \n",
    "#         newformvalue = '%s:%s'%(oldformvalue, adformat)\n",
    "        \n",
    "        # replace old for new fields in the line\n",
    "#         newline = re.sub(oldformat, newformat, line)\n",
    "#         output.write(newformat)\n",
    "        time.sleep(0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "003378e7",
   "metadata": {},
   "outputs": [],
   "source": [
    "f=open('./allpools_gmref_C1F001_wdup.vcf','r').readlines()\n",
    "# print(f[1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "34345af7",
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash\n",
    "grep -v '#' allpools_gmref_C1F001_wdup.vcf | cut -f10  | cut -d':' -f1 | tr -cd '0' | wc -c\n",
    "# 124558098 counts number of zero ocurrences across all genotype fields in the vcf\n",
    "# it's silly to count each possible allele number if I don't know what is the max\n",
    "# removing any character that is not integer then pipe to sort, count uniq\n",
    "\n",
    "# below is for removing all non-digits and keeping line breaks to count alleles in line\n",
    "grep -v '#' allpools_gmref_C1F001_wdup.vcf | cut -f10  | cut -d':' -f1 | tr -cd '[:digit:]\\n' | head\n",
    "\n",
    "# remove duplicate characters per line (results in unique allele digits per line)\n",
    "grep -v '#' allpools_gmref_C1F001_wdup.vcf | cut -f10  | cut -d':' -f1 | tr -cd '[:digit:]\\n' | sed ':1;s/\\(.\\)\\(.*\\)\\1/\\1\\2/;t1' | head\n",
    " \n",
    "# now all data, also sort and count the unique combinations (ordered) of allele counts in the entire vcf\n",
    "grep -v '#' allpools_gmref_C1F001_wdup.vcf | cut -f10  | cut -d':' -f1 | tr -cd '[:digit:]\\n' | sed ':1;s/\\(.\\)\\(.*\\)\\1/\\1\\2/;t1' | sort | uniq -c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a90b8710",
   "metadata": {},
   "outputs": [],
   "source": [
    "# f=open('/home/xxx/scratch/test/' + sys.argv[1],'r').readlines()\n",
    "# print f[1]\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
